File: backend/api/v1/agent.py
"""
API - Agent chat endpoint.

Natural language entry point for the FitAI Agent.
Supports three personas: health_consultant, studio_ops, growth_engine.
"""
import json

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.dependencies import get_current_user
from backend.models.auth import User
from backend.config import settings

router = APIRouter()


class AgentRequest(BaseModel):
    message: str
    member_id: int | None = None
    persona: str | None = None  # "health_consultant" / "studio_ops" / "growth_engine"
    context: dict | None = None


class AgentResponse(BaseModel):
    answer: str
    tool_calls: list[dict]
    iterations: int
    persona: str


@router.post("/chat", response_model=AgentResponse)
async def agent_chat(
    req: AgentRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Agent chat endpoint - natural language interaction with persona support."""
    if not settings.DASHSCOPE_API_KEY:
        raise HTTPException(
            status_code=503,
            detail="Agent service not configured. Set DASHSCOPE_API_KEY in .env",
        )

    from backend.agent.bootstrap import get_orchestrator

    orch = get_orchestrator(db)
    result = await orch.run(
        user_input=req.message,
        user_role=current_user.role if hasattr(current_user, "role") else "front_desk",
        organization_id=current_user.organization_id,
        db=db,
        member_id=req.member_id,
        persona=req.persona,
    )
    return AgentResponse(**result)


@router.get("/tools")
async def list_tools(
    current_user: User = Depends(get_current_user),
):
    """List all available Agent tools."""
    from backend.agent.bootstrap import get_tool_registry
    tools = get_tool_registry()
    return {"tools": tools.list_tools(), "total": len(tools.list_tools())}


@router.get("/health")
async def agent_health():
    """Check if Agent service is configured and ready."""
    return {
        "configured": bool(settings.DASHSCOPE_API_KEY),
        "model": settings.QWEN_MODEL,
        "base_url": settings.LLM_BASE_URL,
        "max_iterations": settings.AGENT_MAX_ITERATIONS,
        "reflection_enabled": settings.AGENT_REFLECTION_ENABLED,
        "personas": ["health_consultant", "studio_ops", "growth_engine"],
        "fallback_enabled": settings.LLM_FALLBACK_ENABLED and bool(settings.LLM_FALLBACK_API_KEY),
        "fallback_model": settings.LLM_FALLBACK_MODEL or None,
    }


@router.get("/personas")
async def list_personas(
    current_user: User = Depends(get_current_user),
):
    """List available Agent personas with descriptions."""
    from backend.agent.personas import AgentRole, PERSONA_TOOLS
    return {
        "personas": [
            {
                "id": AgentRole.HEALTH_CONSULTANT.value,
                "name": "Health Consultant",
                "name_cn": "Health Consultant",
                "description": "Member-facing health advisor - body analysis, course recommendations, attendance tracking",
                "tools": list(PERSONA_TOOLS[AgentRole.HEALTH_CONSULTANT]),
            },
            {
                "id": AgentRole.STUDIO_OPS.value,
                "name": "Studio Ops Assistant",
                "name_cn": "Studio Ops Assistant",
                "description": "Owner-facing operations brain - revenue analysis, coach performance, schedule optimization",
                "tools": list(PERSONA_TOOLS[AgentRole.STUDIO_OPS]),
            },
            {
                "id": AgentRole.GROWTH_ENGINE.value,
                "name": "Growth Engine",
                "name_cn": "Growth Engine",
                "description": "Marketing-facing growth strategist - dormant reactivation, upsell, retention analysis",
                "tools": list(PERSONA_TOOLS[AgentRole.GROWTH_ENGINE]),
            },
        ]
    }


@router.post("/chat/stream")
async def agent_chat_stream(
    req: AgentRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Agent chat with SSE streaming. Returns server-sent events for real-time tool calls and response."""
    if not settings.DASHSCOPE_API_KEY:
        raise HTTPException(
            status_code=503,
            detail="Agent service not configured. Set DASHSCOPE_API_KEY in .env",
        )

    from backend.agent.bootstrap import get_orchestrator

    orch = get_orchestrator(db)

    async def event_generator():
        async for event in orch.run_stream(
            user_input=req.message,
            user_role=current_user.role if hasattr(current_user, "role") else "front_desk",
            organization_id=current_user.organization_id,
            db=db,
            member_id=req.member_id,
            persona=req.persona,
        ):
            yield f"data: {event}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: miniapp/src/services/agent.ts
/**
 * Agent Service - FitAI Agent chat API client.
 * 
 * Provides natural language interaction with the FitAI Agent,
 * supporting three personas: health_consultant, studio_ops, growth_engine.
 */
import { post, get } from './request'

/** Agent persona type */
export type AgentPersona = 'health_consultant' | 'studio_ops' | 'growth_engine'

/** Tool call record */
export interface ToolCall {
  tool: string
  args: Record<string, any>
  result: any
  timestamp: string
}

/** Agent chat response */
export interface AgentChatResponse {
  answer: string
  tool_calls: ToolCall[]
  iterations: number
  persona: string
}

/** Agent chat request */
export interface AgentChatRequest {
  message: string
  member_id?: number
  persona?: AgentPersona
  context?: Record<string, any>
}

/** Persona info */
export interface PersonaInfo {
  id: string
  name: string
  name_cn: string
  description: string
  tools: string[]
}

/**
 * Send a message to the FitAI Agent and get a response.
 */
export async function chatWithAgent(req: AgentChatRequest): Promise<AgentChatResponse> {
  return post<AgentChatResponse>('/api/v1/agent/chat', req, {
    showLoading: false,
    showError: true,
  })
}

/**
 * List all available Agent personas.
 */
export async function getPersonas(): Promise<PersonaInfo[]> {
  const res = await get<{ personas: PersonaInfo[] }>('/api/v1/agent/personas')
  return res.personas
}

/**
 * Check Agent service health.
 */
export async function checkAgentHealth(): Promise<{
  configured: boolean
  model: string
  max_iterations: number
  reflection_enabled: boolean
  personas: string[]
}> {
  return get('/api/v1/agent/health')
}

/**
 * Quick suggestions for each persona.
 * Used to pre-fill the chat input with example prompts.
 */
export const QUICK_SUGGESTIONS: Record<AgentPersona, string[]> = {
  health_consultant: [
    '分析我最近的体测数据趋势',
    '本周有哪些适合我的课程？',
    '我的出勤率怎么样？',
    '帮我预约一节瑜伽课',
  ],
  studio_ops: [
    '本周营业数据概览',
    '教练绩效对比分析',
    '哪些会员超过30天没来了？',
    '本月营收有什么异常？',
  ],
  growth_engine: [
    '设计一个沉睡会员唤醒方案',
    '哪些会员卡即将到期？',
    '找出最有升级潜力的会员',
    '分析高价值会员的消费模式',
  ],
}
